pkg/cmd/roachtest/tests/mixed_version_sql_stats.go

```diff
@@ -0,0 +1,234 @@
+		// SQL Stats are flushed every 10m by default, set it to 3m to speed up the test.
+		l.Printf("setting sql.stats.flush.interval to %s", flushInterval.String())
+		err := h.Exec(r, `SET CLUSTER SETTING sql.stats.flush.interval = $1`, flushInterval.String())
+		return err
+	}
+
+	runGetRequests := func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		// We'll validate that hitting the in-memory tables and either one of the
+		// persisted (full or cached activity) does not return error on any node.
+		// Ideally we'd have prior data in the persisted tables already so that we
+		// can easily query data from the previous aggregation interval. For now
+		// we'll just have to rely on waiting for the flush to occur.
+		expectedTableCount := 2
+
+		// Track the source tables used to service each request.
+		foundStmtTableSources := make(map[string]interface{})
+		foundTxnTableSources := make(map[string]interface{})
+
+		addToTableSources := func(res *serverpb.StatementsResponse) error {
+			if res.StmtsSourceTable != "" {
+				foundStmtTableSources[res.StmtsSourceTable] = struct{}{}
+			}
+			if res.TxnsSourceTable != "" {
+				foundTxnTableSources[res.TxnsSourceTable] = struct{}{}
+			}
+			return nil
+		}
+
+		// These are the two types of stats pages that power the admin UI.
+		requestTypes := []serverpb.CombinedStatementsStatsRequest_StatsType{
+			serverpb.CombinedStatementsStatsRequest_StmtStatsOnly,
+			serverpb.CombinedStatementsStatsRequest_TxnStatsOnly,
+		}
+
+		s := createSQLStatsRequestHelper(c, l)
+		// First, we'll attempt to get stats from the in-memory tables.
+		// We can ensure we request from in-memory tables by requesting stats from
+		// a time range for which no data is available, as the in-memory table
+		// is only used when no data is available from system tables.
+		for _, statsType := range requestTypes {
+			if err := s.requestSQLStatsFromEmptyInterval(ctx, roachNodes, statsType, addToTableSources); err != nil {
+				t.Fatal(err)
+			}
+		}
+		if foundStmtTableSources[server.CrdbInternalStmtStatsCombined] == nil || foundTxnTableSources[server.CrdbInternalTxnStatsCombined] == nil {
+			t.Fatal("expected to find in-memory tables in response, found %v, %v", foundStmtTableSources, foundTxnTableSources)
+		}
+
+		// Now we'll wait for the flush to occur and request stats from the persisted tables.
+		var timer timeutil.Timer
+		defer timer.Stop()
+		timer.Reset(5 * time.Second)
+
+		for len(foundStmtTableSources) < expectedTableCount || len(foundTxnTableSources) < expectedTableCount {
+			select {
+			case <-timer.C:
+				timer.Read = true
+				for _, statsType := range requestTypes {
+					// Requesting data from the last two hours should return data from the persisted tables eventually.
+					if err := s.requestSQLStatsFromLastTwoHours(ctx, roachNodes, statsType, addToTableSources); err != nil {
+						t.Fatal(err)
+					}
+				}
+			case <-ctx.Done():
+				return nil
```

Comment:
Do we want to log something here? In case the test times out, I imagine it would be useful to know the contents of `foundStmtTableSources` and `foundTxnTableSources`?